crates/snapshot/src/segments/mod.rs

```diff
@@ -1,39 +1,68 @@
 //! Snapshot segment implementations and utilities.
 
-mod headers;
+mod transactions;
+pub use transactions::Transactions;
 
+mod headers;
 pub use headers::Headers;
 
-use reth_db::{table::Table, transaction::DbTx};
+mod receipts;
+pub use receipts::Receipts;
+
+use reth_db::{
+    cursor::DbCursorRO, database::Database, table::Table, transaction::DbTx, RawKey, RawTable,
+};
 use reth_interfaces::RethResult;
 use reth_nippy_jar::NippyJar;
 use reth_primitives::{
-    snapshot::{Compression, Filters, InclusionFilter, PerfectHashingFunction},
+    snapshot::{Compression, Filters, InclusionFilter, PerfectHashingFunction, SegmentHeader},
     BlockNumber, SnapshotSegment,
 };
+use reth_provider::{DatabaseProviderRO, TransactionsProviderExt};
 use std::{ops::RangeInclusive, path::PathBuf};
 
 pub(crate) type Rows<const COLUMNS: usize> = [Vec<Vec<u8>>; COLUMNS];
 
 /// A segment represents a snapshotting of some portion of the data.
 pub trait Segment {
     /// Snapshot data using the provided range.
-    fn snapshot(&self, tx: &impl DbTx, range: RangeInclusive<BlockNumber>) -> RethResult<()>;
+    fn snapshot<DB: Database>(
+        &self,
+        provider: &DatabaseProviderRO<'_, DB>,
+        range: RangeInclusive<BlockNumber>,
+    ) -> RethResult<()>;
+
+    /// Generates the dataset to train a zstd dictionary with the most recent rows (at most 1000).
+    fn dataset_for_compression<DB: Database, T: Table<Key = u64>>(
```

Comment:
I'd rather make it a freestanding function, see https://github.com/paradigmxyz/reth/pull/4964#discussion_r1355173722

Reply:
in the future i'd like to have segment specific implementations tho (eg. dont include hashes for headers in the dataset)

Reply:
agree, then trait method makes sense